File: kernel/src/syscalls/access.cpp
#include "process.h"
#include "file_manager.h"

int Process::do_sys_access(UserPtr<const char*> pathname_ptr, int mode) {
	string pathname;
	if (!copy_string_from_user(pathname_ptr, pathname))
		return -EFAULT;
	if (!FileManager::exists(pathname))
		return -EACCES;
	if ((mode & W_OK) || (mode & X_OK)) {
		printf_once("access %s, mode %d, denying\n", pathname.c_str(), mode);
		return -EACCES;
	}
	// It's asking for R_OK or F_OK
	return 0;
}

File: kernel_zig/src/process/syscalls/access.zig
usingnamespace @import("../common.zig");
const mem = @import("../../mem/mem.zig");
const UserCString = mem.safe.UserCString;
const log = std.log.scoped(.sys_access);

fn sys_access(self: *Process, pathname_ptr: UserCString, mode: u32) !void {
    const pathname = try mem.safe.copyStringFromUser(self.allocator, pathname_ptr);
    defer self.allocator.free(pathname);

    // Deny if file doesn't exist, or if asking for W_OK or X_OK. We only
    // allow R_OK and F_OK
    if (!fs.file_manager.exists(pathname))
        return error.PermissionDenied;
    if ((mode & linux.W_OK) or (mode & linux.X_OK)) {
        log.warn("denying {}, mode {}\n", .{ pathname, mode });
        return error.PermissionDenied;
    }
}

pub fn handle_sys_access(self: *Process, arg0: usize, arg1: usize) !usize {
    const pathname_ptr = UserCString.fromFlat(arg0);
    const mode = std.meta.cast(u32, arg1);
    try sys_access(self, pathname_ptr, mode);
    return 0;
}
